rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,171 @@
+pub struct SqlBuilder {
+    /// The dataset to run the SQL query
+    pub(crate) dataset: Option<Dataset>,
+
+    /// The SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: Option<String>,
+
+    /// If true, the query result will include the internal row id
+    pub(crate) row_id: bool,
+
+    /// If true, the query result will include the internal row address
+    pub(crate) row_addr: bool,
+}
+
+impl SqlBuilder {
+    /// The table name to register in the datafusion context.
+    /// This is used to specify a "table name" for the dataset.
+    /// So that you can run SQL queries against it.
+    /// If not set, the default table name is "dataset".
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = Some(table_name.to_string());
+        self
+    }
+
+    /// Specify if the query result should include the internal row id.
+    /// If true, the query result will include an additional column named "_rowid".
+    pub fn with_row_id(mut self, row_id: bool) -> Self {
+        self.row_id = row_id;
+        self
+    }
+
+    /// Specify if the query result should include the internal row address.
+    /// If true, the query result will include an additional column named "_rowaddr".
+    pub fn with_row_addr(mut self, row_addr: bool) -> Self {
+        self.row_addr = row_addr;
+        self
+    }
+
+    pub async fn execute(self) -> lance_core::Result<QueryResult> {
+        let ctx = SessionContext::new();
+        ctx.register_table(
+            self.table_name.unwrap(),
+            Arc::new(LanceTableProvider::new(
+                Arc::new(self.dataset.unwrap()),
+                self.row_id,
+                self.row_addr,
+            )),
+        )?;
+        let df = ctx.sql(&self.sql).await?;
+        Ok(QueryResult::new(df))
+    }
+}
+
+impl Default for SqlBuilder {
+    fn default() -> Self {
+        Self {
+            dataset: None,
+            sql: "".to_string(),
+            table_name: Some("dataset".to_string()),
+            row_id: false,
+            row_addr: false,
+        }
+    }
+}
+
+pub struct QueryResult {
+    dataframe: DataFrame,
+}
+
+impl QueryResult {
+    pub fn new(dataframe: DataFrame) -> Self {
+        Self { dataframe }
+    }
+
+    pub async fn into_stream(self) -> SendableRecordBatchStream {
+        self.dataframe.execute_stream().await.unwrap()
+    }
+
+    pub async fn collect(self) -> lance_core::Result<Vec<RecordBatch>> {
+        use futures::TryStreamExt;
+        Ok(self
+            .dataframe
+            .execute_stream()
+            .await
+            .unwrap()
+            .try_collect::<Vec<_>>()
+            .await?)
+    }
+
+    pub fn dataframe(&self) -> &DataFrame {
+        &self.dataframe
+    }
+
+    pub async fn explain(&self, verbose: bool, analyze: bool) -> DataFrame {
+        self.dataframe.clone().explain(verbose, analyze).unwrap()
```

Comment:
I think based on what we discussed before, we don't want to expose the dataframe itself. For explain, the returned DataFrame has one string column named "plan". So I think we can do something like

```rust
    let explained_df = df.explain(false, false)?;
    let batches = explained_df.collect().await?;
    let mut lines = Vec::new();
    for batch in &batches {
        let column = batch.column(0);
        let array = column
            .as_any()
            .downcast_ref::<StringArray>()
            .expect("Expected StringArray in 'plan' column for DataFrame.explain");
        for i in 0..array.len() {
            lines.push(array.value(i).to_string());
        }
    }

    Ok(lines.join("\n"))
```